vendor/engines/bulk_email/app/controllers/bulk_email/bulk_email_controller.rb

```diff
@@ -11,26 +11,50 @@ class BulkEmailController < ApplicationController
     before_action :authenticate_user!
     before_action :check_acting_as
     before_action :init_current_facility
+    before_action :init_delivery_form, only: [:create, :deliver]
     before_action { authorize! :send_bulk_emails, current_facility }
 
-    before_action :init_search_options
+    before_action :init_search_options, only: [:search]
 
     helper_method :datepicker_field_input
     helper_method :user_type_selected?
 
     def search
-      @searcher = BulkEmailSearcher.new(@search_fields)
+      @searcher = RecipientSearcher.new(@search_fields)
       @users = @searcher.do_search
     end
 
-    def download
+    def create
       @users = User.where_ids_in(params[:recipient_ids])
-      filename = "bulk_email_recipients.csv"
-      set_csv_headers(filename)
+
+      respond_to do |format|
+        format.csv do
+          filename = "bulk_email_recipients.csv"
+          set_csv_headers(filename)
+        end
+
+        format.html
+      end
+    end
+
+    def deliver
+      if @delivery_form.valid?
+        @delivery_form.deliver_all!
+        flash[:notice] = text("bulk_email.delivery.success", count: @delivery_form.recipient_ids.count)
+        redirect_to facility_bulk_email_path
+      else
+        @users = User.where_ids_in(@delivery_form.recipient_ids)
+        render :create
+      end
     end
 
     private
 
+    def init_delivery_form
+      @delivery_form = DeliveryForm.new(current_facility)
+      @delivery_form.assign_attributes(params) if params[:bulk_email_delivery_form].present?
```

Comment:
Good catch: it was like this because an earlier version did not have `recipient_ids` nested within `:bulk_email_delivery_form`. It's there now, so there's no longer a need to send all of `params`.